kafka/producer/base.py

```diff
@@ -200,7 +253,16 @@ def _send_messages(self, topic, partition, *msg, **kwargs):
 
         if self.async:
             for m in msg:
-                self.queue.put((TopicAndPartition(topic, partition), m, key))
+                try:
+                    item = (TopicAndPartition(topic, partition), m, key)
+                    if self.async_queue_put_timeout == 0:
+                        self.queue.put_nowait(item)
+                    else:
+                        self.queue.put(item, True, self.async_queue_put_timeout)
+                except Full:
+                    raise AsyncProducerQueueFull(
```

Comment:
Good catch! My proposal is tracking index `n` and return `msg[n+1:]` as `failed_msgs` property's value of `AsyncProducerQueueFull` exception.